internal/venice-client-common/src/main/java/org/apache/avro/generic/DeterministicMapOrderDatumWriter.java

```diff
@@ -14,20 +16,46 @@
  * the same as natural ordering of map keys.
  */
 public interface DeterministicMapOrderDatumWriter {
+  Comparator<Map.Entry<? extends CharSequence, Object>> COMPARATOR = (e1, e2) -> {
+    // TODO: Replace with CharSequence#compare once the code has completely migrated to JDK11+
+    CharSequence cs1 = e1.getKey();
+    CharSequence cs2 = e2.getKey();
+    if (Objects.requireNonNull(cs1) == Objects.requireNonNull(cs2)) {
+      return 0;
+    }
+
+    if (cs1.getClass() == cs2.getClass() && cs1 instanceof Comparable) {
+      return ((Comparable<Object>) cs1).compareTo(cs2);
+    }
+
+    for (int i = 0, len = Math.min(cs1.length(), cs2.length()); i < len; i++) {
+      char a = cs1.charAt(i);
+      char b = cs2.charAt(i);
+      if (a != b) {
+        return a - b;
+      }
+    }
+
+    return cs1.length() - cs2.length();
+  };
+
   void internalWrite(Schema schema, Object datum, Encoder out) throws IOException;
 
   default void writeMapWithDeterministicOrder(Schema schema, Object datum, Encoder out) throws IOException {
     Schema valueSchemaType = schema.getValueType();
-    Map map = (Map) datum;
+
+    @SuppressWarnings("unchecked")
+    Map<? extends CharSequence, Object> map = (Map<? extends CharSequence, Object>) datum;
```

Comment:
`Map with generic type cast` doesn't throw exception if the map key class is not `CharSequence`.
So shall we did a type check in the `COMPARATOR` and throw exception if the key class is not `CharSequence`?

Reply:
I see. In that case, yeah, we should throw an exception

Reply:
Looks like `GenericDatumWriter` in Avro doesn't care if the key is a `String` type or not. It only calls `.toString()` on the map key directly ([Link](https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/generic/GenericDatumWriter.java#L339)). I'll update the code to follow the same semantics